### ISSUE TITLE
How to convert from PerspectiveCamera object to pulsar renderer camera parameters?

### ISSUE THREAD

**Issue by weders** [facebookresearch/pytorch3d#734]
## Camera parameter conversion

I have a PerspectiveCamera object that works with a point cloud of a scene and the PointsRenderer. I would like to use the Pulsar renderer as defined in 
https://github.com/facebookresearch/pytorch3d/blob/master/pytorch3d/renderer/points/pulsar/renderer.py

Unfortunately, it is not clear to me how to convert the camera parameters of the PerspectiveCamera object to the pulsar format, which is defined as [position, rotation, focal_length, sensor_size, principal_point].

I have written the following method

```
def transform_camera_to_pulsar(camera):

        position = cameras.get_camera_center()
        angles = matrix_to_euler_angles(cameras.get_world_to_view_transform().get_matrix()[:, :3, :3], convention='XYZ')
        focal_length = cameras.focal_length[:, 0].unsqueeze(-1)
        principal_point = cameras.principal_point
        sensor_size = torch.ones_like(focal_length)


        camera = torch.cat([position, angles, focal_length, sensor_size, principal_point], dim=-1).cpu()
        camera = camera[0]
        
        return camera
```

However, this does not work at all. Is the pulsar renderer using another camera convention? Or do I miss something else?

**Comment by patricklabatut** [comment]
Cameras for the Pulsar renderer do follow a different convention and have some constraints (e.g. same focal lengths on both axis, left-handed coordinate system, vertical flip).

We are however working on adding convenience conversion functions.

**Comment by classner** [comment]
Hi @weders ,

I have already implemented a function for this and am just waiting for approval to land it. I'll keep you posted here once it's there (should be in the next days).

**Comment by classner** [comment]
Functionality implemented with https://github.com/facebookresearch/pytorch3d/commit/75432a069503a2a51a79516f10ae8eb246e816d0 . You can now

```
from pytorch3d.utils import pulsar_from_cameras_projection

# ...
pulsar_params = pulsar_from_cameras_projection(cameras, image_size)
```

Let me know if you have any further questions!

**Comment by weders** [comment]
Awesome! Thanks a lot to making this work!

**Comment by xuelin-chen** [comment]
> Cameras for the Pulsar renderer do follow a different convention and have some constraints (e.g. same focal lengths on both axis, left-handed coordinate system, vertical flip).
> 
> We are however working on adding convenience conversion functions.

hi, @patricklabatut thanks for the excellent codes, this may be a stupid question,: what has stopped it from using different focal lengths? is there any possibility or plan to suppor that? thank you!:-)

**Comment by classner** [comment]
Hi @xuelin-chen,

I implemented Pulsar originally with a pure pinhole model in mind, assuming undistorted and standardized image. We then moved further and further into classical vision applications working with cameras with lens models increasingly far away from the pinhole model. That’s why I added the center point later, but never got around to introduce separate focal length values. For nearly all lenses in correct calibration models, the focal length in x and y direction is very close to the same value, so taking the average of the two is a good solution (and it is wise to check that the two averaged values are close to each other - otherwise it is worthwhile revisiting the calibration). Changing it is also not a big problem: I was pretty consistent with variable naming, and this [search query](https://github.com/search?l=&p=1&q=focal_length+repo%3Afacebookresearch%2Fpytorch3d+path%3Apytorch3d%2Frenderer%2Fpoints%2Fpulsar%2F+path%3Apytorch3d%2Fcsrc%2Fpulsar%2F&ref=advsearch&type=Code) should give you a good starting point. PRs welcome!

**Comment by xuelin-chen** [comment]
Hi @classner ,

Thanks so much for your so detailed reply.
I see now, it is indeed that the focal lengths are very close usually. Let me see if this tiny error from the focal length average will or not cause a storm in my code before we have to change it and do a PR. ;-)
Thank you!

**Comment by saulzar** [comment]
Hi @xuelin-chen @classner 

The other thing I believe you can do is to resize your images to remove the aspect ratio - I have been doing this with OpenCV with `initUndistortRectifyMap` by scaling the second row of the camera matrix (K) to make fx = fy.

This seems to work fine for aspect ratio differences, as far as I can see? My calibration previously also had skew - and `initUndistortRectifyMap` doesn't work with that at all!

Cheers,
Oliver
